rinex/src/epoch/mod.rs

```diff
@@ -122,84 +129,88 @@ pub(crate) fn format(epoch: Epoch, flag: Option<EpochFlag>, t: Type, revision: u
 }
 
 /*
- * Parses an Epoch and optional flag, from standard specifications.
- * YY encoded on two digits, prior 20000 get shifted to 21st century.
+ * Parses an Epoch and optional flag, interpreted as a datetime within specified TimeScale.
  */
-pub(crate) fn parse(s: &str) -> Result<(Epoch, EpochFlag), Error> {
+pub(crate) fn parse_in_timescale(
+    s: &str,
+    ts: TimeScale,
+) -> Result<(Epoch, EpochFlag), ParsingError> {
     let items: Vec<&str> = s.split_ascii_whitespace().collect();
     if items.len() != 6 && items.len() != 7 {
-        return Err(Error::FormatError);
+        return Err(ParsingError::FormatError);
     }
-    if let Ok(mut y) = i32::from_str_radix(items[0], 10) {
-        if y < 100 {
-            // two digit issues (old rinex format)
-            if y < 80 {
-                // RINEX did not exist
-                // modern file (2000+) that uses old revision,
-                y += 2000;
-            } else {
-                y += 1900; // [1980:2000]
-            }
+
+    let mut y = i32::from_str_radix(items[0], 10)
```

Comment:
Not sure how much it matters, but you can repeatedly call `Iterator::next` instead of `collect`, in order to avoid the allocation for `items`.

Reply:
Oh you mean how we manage `&str.split_ascii_whitespace()` I presume.

We have that in several locations and it could be significant if they get all improved.

It's just I don't a "clean" way to deal for example in this case, when parsing a datetime when you need to collect 6 or 7 fields, I find using `iter.next()` is very verbose. For example : 

```rust
let y_str = iter.next()
    .unwrap();
let y = u32::parse();

let m_str = iter.next()
   .unwrap();
let m = u32::parse();
```

to reduce the code quantity we should have something like `while let Some(str) = iter.next()`  but we need a robust way to identify which field we're about to parse

Reply:
I managed to get it to work :) I
it's not that neat, because I have a custom errors like `Error::YearField` or `Error::MonthField` that makes debugging very easy, that I think we should preserve. Considering how RINEX differs in its format.

I'll see in the next few weeks, if we can improve other spots where I used `split.ascii_whitespace().collect()`

Reply:
Right. But you can still do something like:

```rust
    let mut it = content.split_ascii_whitespace();

    let item = it
        .next()
        .ok_or_else(|| ParsingError::YearField("missing year".to_string()))?;
    let mut y = item
        .parse::<i32>()
        .map_err(|_| ParsingError::YearField(item.to_string()))?;
    /* old RINEX problem: YY is sometimes encoded on two digits */
    if y < 100 {
        if y < 80 {
            y += 2000;
        } else {
            y += 1900;
        }
    }

    let item = it
        .next()
        .ok_or_else(|| ParsingError::MonthField("missing month".to_string()))?;
    let month = item
        .parse::<u8>()
        .map_err(|_| ParsingError::MonthField(item.to_string()))?;
```

Or, with a small helper, a variation like:

```rust
    fn parse_number<'a, T: FromStr<Err = FSE>, FSE, E>(
        s: &'a str,
        invalid: impl FnOnce(&'a str) -> E,
    ) -> Result<T, E> {
        s.parse::<T>().map_err(|_| invalid(s))
    }

    let mut it = content.split_ascii_whitespace();

    let mut y = parse_number(
        it.next()
            .ok_or_else(|| ParsingError::YearField("missing Year".to_string()))?,
        |s| ParsingError::YearField(s.to_string()),
    )?;
    /* old RINEX problem: YY is sometimes encoded on two digits */
    if y < 100 {
        if y < 80 {
            y += 2000;
        } else {
            y += 1900;
        }
    }
    let mut m = parse_number(
        it.next()
            .ok_or_else(|| ParsingError::MonthField("missing month".to_string()))?,
        |s| ParsingError::MonthField(s.to_string()),
    )?;
```